<a href="https://www.kaggle.com/code/kiranpatil7022/seoul-bike-insights-and-predictions?scriptVersionId=147390888" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

## **1. Loading libraries and data**

In [1]:
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/seoul-bike-trip-duration-prediction/For_modeling.csv


In [2]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import  train_test_split

In [3]:
df = pd.read_csv(
    "/kaggle/input/seoul-bike-trip-duration-prediction/For_modeling.csv",
    dtype={
        'Duration' : 'int8',
        'Distance' : 'int8',
        'PLong' : 'float32',
        'PLatd' : 'float32',
        'DLong' : 'float32',
        'DLatd' : 'float32',
        'Haversine' : 'float32',
        'Pmonth' : 'int8',
        'Pday' : 'int8',
        'Phour' : 'int8',
        'Pmin' : 'int8',
        'PDweek' : 'int8',
        'Dmonth' : 'int8',
        'Dday' : 'int8',
        'Dhour' : 'int8',
        'Dmin' : 'int8',
        'DDweek' : 'int8',
        'Temp' : 'float32',
        'Precip' : 'float32',
        'Wind' : 'float32',
        'Humid' : 'float32',
        'Solar' : 'float32',
        'Snow' : 'float32',
        'GroundTemp' : 'float32',
        'Dust' : 'float32'
    },
    index_col=0
).sample(frac=1)
df.head()

Duration  Distance      PLong       PLatd      DLong       DLatd  \
7351211        60        42  37.518970  126.869965  37.518284  126.912636   
8988067         4       -58  37.563858  127.030319  37.561096  127.036797   
6756042        56       -20  37.545284  126.931053  37.524071  127.021790   
9058257       101       104  37.574718  127.053391  37.551250  127.035103   
4333197        26      -112  37.567966  126.931725  37.567966  126.931725   

         Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek       Temp  Precip  \
7351211   3.764141      10     8     13  ...    59       0  20.200001     0.0   
8988067   0.648347      11    18      0  ...    46       6   3.600000     0.0   
6756042   8.341287       9    26     18  ...     0       2  22.100000     0.0   
9058257   3.067248      11    20      9  ...    31       1   2.300000     0.0   
4333197   0.000000       7    24     13  ...    42       1  36.400002     0.0   

         Wind  Humid  Solar  Snow  GroundTemp  Dust  
7351211   1.3   28.0   2.25   0.0   25.400000  24.0  
8988067   0.8   49.0   0.00   0.0    0.700000  29.0  
6756042   1.5   45.0   0.27   0.0   21.900000  12.0  
9058257   1.2   43.0   0.52   0.0   -0.100000  31.0  
4333197   1.3   41.0   2.86   0.0   57.099998  24.0  

[5 rows x 25 columns]

In [4]:
df.shape

(9601139, 25)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9601139 entries, 7351211 to 9043268
Data columns (total 25 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Duration    int8   
 1   Distance    int8   
 2   PLong       float32
 3   PLatd       float32
 4   DLong       float32
 5   DLatd       float32
 6   Haversine   float32
 7   Pmonth      int8   
 8   Pday        int8   
 9   Phour       int8   
 10  Pmin        int8   
 11  PDweek      int8   
 12  Dmonth      int8   
 13  Dday        int8   
 14  Dhour       int8   
 15  Dmin        int8   
 16  DDweek      int8   
 17  Temp        float32
 18  Precip      float32
 19  Wind        float32
 20  Humid       float32
 21  Solar       float32
 22  Snow        float32
 23  GroundTemp  float32
 24  Dust        float32
dtypes: float32(13), int8(12)
memory usage: 659.3 MB


## **2. Exploratory Data Analysis**

In [6]:
df.columns

Index(['Duration', 'Distance', 'PLong', 'PLatd', 'DLong', 'DLatd', 'Haversine',
       'Pmonth', 'Pday', 'Phour', 'Pmin', 'PDweek', 'Dmonth', 'Dday', 'Dhour',
       'Dmin', 'DDweek', 'Temp', 'Precip', 'Wind', 'Humid', 'Solar', 'Snow',
       'GroundTemp', 'Dust'],
      dtype='object')

In [7]:
df.describe()

Duration      Distance         PLong         PLatd         DLong  \
count  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   2.580294e+01 -9.069010e-01  3.754741e+01  1.269908e+02  3.754747e+01   
std    2.504333e+01  7.404757e+01  4.441376e-02  8.257065e-02  4.445305e-02   
min    1.000000e+00 -1.280000e+02  3.743727e+01  1.267986e+02  3.743727e+01   
25%    8.000000e+00 -6.600000e+01  3.751424e+01  1.269200e+02  3.751424e+01   
50%    1.600000e+01  0.000000e+00  3.754693e+01  1.269943e+02  3.754655e+01   
75%    3.600000e+01  6.400000e+01  3.757324e+01  1.270618e+02  3.757324e+01   
max    1.190000e+02  1.260000e+02  3.768972e+01  1.271803e+02  3.768972e+01   

              DLatd     Haversine        Pmonth          Pday         Phour  \
count  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   1.269908e+02  1.870315e+00  7.578754e+00  1.576714e+01  1.437801e+01   
std    8.323726e-02  1.995361e+00  2.696797e+00  8.748012e+00  6.249244e+00   
min    1.267986e+02  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    1.269194e+02  6.758669e-01  6.000000e+00  8.000000e+00  1.000000e+01   
50%    1.269947e+02  1.257498e+00  8.000000e+00  1.600000e+01  1.600000e+01   
75%    1.270624e+02  2.363812e+00  1.000000e+01  2.300000e+01  1.900000e+01   
max    1.271803e+02  2.863448e+01  1.200000e+01  3.100000e+01  2.300000e+01   

       ...          Dmin        DDweek          Temp        Precip  \
count  ...  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   ...  2.999278e+01  2.945771e+00  1.856201e+01  1.784855e-02   
std    ...  1.730960e+01  1.977624e+00  9.498075e+00  3.736281e-01   
min    ...  0.000000e+00  0.000000e+00 -1.780000e+01  0.000000e+00   
25%    ...  1.500000e+01  1.000000e+00  1.220000e+01  0.000000e+00   
50%    ...  3.000000e+01  3.000000e+00  2.010000e+01  0.000000e+00   
75%    ...  4.500000e+01  5.000000e+00  2.550000e+01  0.000000e+00   
max    ...  5.900000e+01  6.000000e+00  3.940000e+01  3.500000e+01   

               Wind         Humid         Solar          Snow    GroundTemp  \
count  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   1.819273e+00  5.397557e+01  7.300679e-01  9.948919e-03  2.072376e+01   
std    9.710258e-01  1.742159e+01  9.175004e-01  1.628091e-01  1.265221e+01   
min    0.000000e+00  1.000000e+01  0.000000e+00  0.000000e+00 -1.360000e+01   
25%    1.100000e+00  4.100000e+01  0.000000e+00  0.000000e+00  1.160000e+01   
50%    1.700000e+00  5.300000e+01  2.700000e-01  0.000000e+00  2.110000e+01   
75%    2.400000e+00  6.700000e+01  1.260000e+00  0.000000e+00  2.850000e+01   
max    7.400000e+00  9.800000e+01  3.520000e+00  8.800000e+00  6.220000e+01   

               Dust  
count  9.601139e+06  
mean   3.245800e+01  
std    2.407264e+01  
min    0.000000e+00  
25%    1.600000e+01  
50%    2.700000e+01  
75%    4.200000e+01  
max    3.040000e+02  

[8 rows x 25 columns]

**Observations:**
1. Distance: there are data points where the distance value is negatice (-ve)
2. Haversine: there are data points where the haversine value is 0. 

**Observation analysis 1. Distance**  
Exploring the data points where the distance is negative.

In [8]:
df.shape

(9601139, 25)

In [9]:
df[df['Distance']<0].shape

(4787174, 25)

In [10]:
df[df['Distance']<0].head()

Duration  Distance      PLong       PLatd      DLong       DLatd  \
8988067         4       -58  37.563858  127.030319  37.561096  127.036797   
6756042        56       -20  37.545284  126.931053  37.524071  127.021790   
4333197        26      -112  37.567966  126.931725  37.567966  126.931725   
9075094        40       -54  37.536430  126.871521  37.529060  126.848480   
536041         26      -128  37.524681  127.135406  37.541805  127.124718   

         Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek       Temp  Precip  \
8988067   0.648347      11    18      0  ...    46       6   3.600000     0.0   
6756042   8.341287       9    26     18  ...     0       2  22.100000     0.0   
4333197   0.000000       7    24     13  ...    42       1  36.400002     0.0   
9075094   2.190768      11    20     19  ...    19       1   8.700000     0.0   
536041    2.124574       3    17     14  ...     2       5  12.100000     0.0   

         Wind  Humid  Solar  Snow  GroundTemp  Dust  
8988067   0.8   49.0   0.00   0.0    0.700000  29.0  
6756042   1.5   45.0   0.27   0.0   21.900000  12.0  
4333197   1.3   41.0   2.86   0.0   57.099998  24.0  
9075094   2.2   34.0   0.00   0.0    4.800000  38.0  
536041    2.3   29.0   2.26   0.0   20.400000  47.0  

[5 rows x 25 columns]

In [11]:
# transforming all the -ve distance to +ve distance
df['Distance'] = df['Distance'].apply(lambda x: abs(x))

In [12]:
df[df['Distance']<0].shape

(0, 25)

**Observation analysis 2. haversine**    
Exploring the data points where the haversine value is 0.   
It should mean that pick up and drop off location (longitude and latitude) are the same.

In [13]:
df.shape

(9601139, 25)

In [14]:
df[df['Haversine']==0].shape

(955083, 25)

In [15]:
df[df['Haversine']==0].reset_index().drop(columns=['index']).head()

Duration  Distance      PLong       PLatd      DLong       DLatd  \
0        26       112  37.567966  126.931725  37.567966  126.931725   
1       113        72  37.514240  127.123070  37.514240  127.123070   
2        59        38  37.544582  127.044609  37.544582  127.044609   
3        50        58  37.600700  126.920128  37.600700  126.920128   
4         9        40  37.577793  127.057831  37.577793  127.057831   

   Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek       Temp  Precip  Wind  \
0        0.0       7    24     13  ...    42       1  36.400002     0.0   1.3   
1        0.0       8    18     17  ...    14       5  29.900000     0.0   2.9   
2        0.0       7     8     17  ...    39       6  27.400000     0.0   1.7   
3        0.0       4     8     20  ...    51       6   2.700000     0.0   1.5   
4        0.0       9     8     23  ...    24       5  20.700001     0.0   1.4   

   Humid  Solar  Snow  GroundTemp  Dust  
0   41.0   2.86   0.0   57.099998  24.0  
1   39.0   0.86   0.0   40.000000  22.0  
2   44.0   1.17   0.0   36.799999  12.0  
3   90.0   0.00   0.0    4.400000   4.0  
4   52.0   0.00   0.0   18.900000  30.0  

[5 rows x 25 columns]

In [16]:
df[df['Haversine']==0].reset_index().drop(columns=['index']).describe()

Duration       Distance          PLong          PLatd  \
count  955083.000000  955083.000000  955083.000000  955083.000000   
mean       49.878425      63.555497      37.546593     126.989716   
std        30.473000      36.989730       0.045334       0.084399   
min         2.000000       0.000000      37.437271     126.798599   
25%        26.000000      32.000000      37.514088     126.918503   
50%        47.000000      64.000000      37.542816     126.995476   
75%        69.000000      96.000000      37.572113     127.062424   
max       119.000000     128.000000      37.689720     127.180267   

               DLong          DLatd  Haversine         Pmonth           Pday  \
count  955083.000000  955083.000000   955083.0  955083.000000  955083.000000   
mean       37.546593     126.989716        0.0       7.367401      15.648030   
std         0.045334       0.084399        0.0       2.556475       8.761283   
min        37.437271     126.798599        0.0       1.000000       1.000000   
25%        37.514088     126.918503        0.0       5.000000       8.000000   
50%        37.542816     126.995476        0.0       8.000000      16.000000   
75%        37.572113     127.062424        0.0       9.000000      23.000000   
max        37.689720     127.180267        0.0      12.000000      31.000000   

               Phour  ...           Dmin         DDweek           Temp  \
count  955083.000000  ...  955083.000000  955083.000000  955083.000000   
mean       15.695371  ...      29.606122       3.191605      20.290329   
std         5.232936  ...      17.347721       2.069881       8.441665   
min         0.000000  ...       0.000000       0.000000     -17.799999   
25%        13.000000  ...      15.000000       1.000000      15.700000   
50%        17.000000  ...      29.000000       3.000000      21.600000   
75%        20.000000  ...      45.000000       5.000000      26.299999   
max        23.000000  ...      59.000000       6.000000      39.400002   

              Precip           Wind          Humid          Solar  \
count  955083.000000  955083.000000  955083.000000  955083.000000   
mean        0.015202       1.937013      51.137737       0.896492   
std         0.309063       0.961516      17.037958       0.991830   
min         0.000000       0.000000      10.000000       0.000000   
25%         0.000000       1.200000      38.000000       0.000000   
50%         0.000000       1.800000      50.000000       0.510000   
75%         0.000000       2.600000      63.000000       1.680000   
max        35.000000       7.400000      98.000000       3.520000   

                Snow     GroundTemp           Dust  
count  955083.000000  955083.000000  955083.000000  
mean        0.007591      23.291319      32.303020  
std         0.143294      11.826364      23.582991  
min         0.000000     -13.600000       0.000000  
25%         0.000000      15.800000      16.000000  
50%         0.000000      23.299999      27.000000  
75%         0.000000      30.299999      42.000000  
max         8.800000      62.200001     304.000000  

[8 rows x 25 columns]

- Despite having a haversine distance of 0, indicating that the pick-up and drop-off locations are identical, the duration of trips exhibits a minimum value of 2 minutes and a maximum value of 119 minutes. 
- This suggests that bicycles may have traveled round-trip journeys, returning to the pick-up location after visiting other destinations. 
- The removal of such data points could potentially impact the performance of the model.

In [17]:
df[df['Distance']==0].shape

(69022, 25)

In [18]:
df[df['Distance']==0].head()

Duration  Distance      PLong       PLatd      DLong       DLatd  \
9534781         4         0  37.538052  127.076599  37.542519  127.084084   
6320088        12         0  37.491131  127.125809  37.508984  127.126595   
3997555        11         0  37.530148  126.968483  37.534840  126.977661   
6316875        56         0  37.556610  126.898018  37.610779  126.917351   
5085378         6         0  37.518738  126.895576  37.522816  126.885651   

         Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek       Temp  Precip  \
9534781   0.825984      12    12      2  ...    16       2  -2.200000     0.0   
6320088   1.986376       9    16      1  ...    13       6  22.799999     0.0   
3997555   0.962897       7    15     13  ...    40       6  31.400000     0.0   
6316875   6.259606       9    15     22  ...    49       5  23.100000     0.0   
5085378   0.985793       8    16     21  ...     9       3  26.799999     0.0   

         Wind  Humid  Solar  Snow  GroundTemp  Dust  
9534781   1.6   72.0   0.00   0.0   -3.500000  77.0  
6320088   0.9   84.0   0.00   0.0   22.299999  35.0  
3997555   3.2   53.0   3.28   0.0   48.500000  13.0  
6316875   0.6   79.0   0.00   0.0   22.200001  15.0  
5085378   2.0   46.0   0.00   0.0   26.799999  18.0  

[5 rows x 25 columns]

In [19]:
filtered_df = df.loc[(df['Distance'] == 0) &
                      (df['Haversine'] == 0) &
                      (df['Pmonth'] == df['Dmonth']) &
                      (df['Pday'] == df['Dday']) &
                      (df['Phour'] == df['Dhour']) &
                      (df['Pmin'] == df['Dmin']) &
                      (df['PDweek'] == df['DDweek'])]

# Print the filtered DataFrame
filtered_df

Duration  Distance      PLong       PLatd      DLong       DLatd  \
1105116        42         0  37.483990  127.034180  37.483990  127.034180   
7146901        54         0  37.520451  127.104202  37.520451  127.104202   
7146902        54         0  37.520451  127.104202  37.520451  127.104202   
1105117        42         0  37.483990  127.034180  37.483990  127.034180   
7410415        65         0  37.526264  126.920509  37.526264  126.920509   
7417862        97         0  37.523300  127.038475  37.523300  127.038475   
7410565        31         0  37.573666  127.030815  37.573666  127.030815   
7410413        65         0  37.526264  126.920509  37.526264  126.920509   
7417860        97         0  37.523300  127.038475  37.523300  127.038475   
7410409        65         0  37.526264  126.920509  37.526264  126.920509   
7146899        54         0  37.520451  127.104202  37.520451  127.104202   
7417857        97         0  37.523300  127.038475  37.523300  127.038475   
7410564        31         0  37.573666  127.030815  37.573666  127.030815   
7410566        31         0  37.573666  127.030815  37.573666  127.030815   
1105115        42         0  37.483990  127.034180  37.483990  127.034180   

         Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek       Temp  Precip  \
1105116        0.0       4    17     19  ...     7       1  14.500000     0.0   
7146901        0.0      10     3     17  ...    10       2  23.299999     0.0   
7146902        0.0      10     3     17  ...    10       2  23.299999     0.0   
1105117        0.0       4    17     19  ...     7       1  14.500000     0.0   
7410415        0.0      10     9     16  ...    18       1  19.000000     0.0   
7417862        0.0      10     9     16  ...    29       1  19.000000     0.0   
7410565        0.0      10     9     16  ...    22       1  19.000000     0.0   
7410413        0.0      10     9     16  ...    18       1  19.000000     0.0   
7417860        0.0      10     9     16  ...    29       1  19.000000     0.0   
7410409        0.0      10     9     16  ...    18       1  19.000000     0.0   
7146899        0.0      10     3     17  ...    10       2  23.299999     0.0   
7417857        0.0      10     9     16  ...    29       1  19.000000     0.0   
7410564        0.0      10     9     16  ...    22       1  19.000000     0.0   
7410566        0.0      10     9     16  ...    22       1  19.000000     0.0   
1105115        0.0       4    17     19  ...     7       1  14.500000     0.0   

         Wind  Humid  Solar  Snow  GroundTemp  Dust  
1105116   3.1   31.0   0.25   0.0   13.200000  91.0  
7146901   1.7   36.0   0.95   0.0   25.299999  19.0  
7146902   1.7   36.0   0.95   0.0   25.299999  19.0  
1105117   3.1   31.0   0.25   0.0   13.200000  91.0  
7410415   1.4   39.0   0.89   0.0   19.600000  28.0  
7417862   1.4   39.0   0.89   0.0   19.600000  28.0  
7410565   1.4   39.0   0.89   0.0   19.600000  28.0  
7410413   1.4   39.0   0.89   0.0   19.600000  28.0  
7417860   1.4   39.0   0.89   0.0   19.600000  28.0  
7410409   1.4   39.0   0.89   0.0   19.600000  28.0  
7146899   1.7   36.0   0.95   0.0   25.299999  19.0  
7417857   1.4   39.0   0.89   0.0   19.600000  28.0  
7410564   1.4   39.0   0.89   0.0   19.600000  28.0  
7410566   1.4   39.0   0.89   0.0   19.600000  28.0  
1105115   3.1   31.0   0.25   0.0   13.200000  91.0  

[15 rows x 25 columns]

In [20]:
filtered_df.shape

(15, 25)